### PR TITLE
Rewrite assert statement with torch._assert under config

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88246

This diff rewrites assert statement in python with torch._assert under config. The resulting graph looks something like:
```
SOURCE CODE: 
def f(x):
      assert x[0] == 3
      return x.cos()
 
CAPTURED GRAPH:   
graph():
    %arg0 : [#users=2] = placeholder[target=arg0]
    %getitem : [#users=1] = call_function[target=operator.getitem](args = (%arg0, 0), kwargs = {})
    %eq : [#users=1] = call_function[target=operator.eq](args = (%getitem, 3), kwargs = {})
    %_assert : [#users=0] = call_function[target=torch._assert](args = (%eq, "assertion_error"), kwargs = {})
    %cos : [#users=1] = call_method[target=cos](args = (%arg0,), kwargs = {})
    return cos
 ```
Note that this introduces side-effect as it could error out while executing graph, but the assertion can eliminated via DCE if we choose to ignore it. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -1758,6 +1758,36 @@ def forward(self, inp):
         args = (torch.randn(3, 4),)
         self.assertTrue(same(mod(*args), opt_mod(*args)))
 
+    @patch.object(torch._dynamo.config, "rewrite_assert_with_torch_assert", True)
+    def test_rewrite_assert(self):
+        def f(x):
+            b = x.sin()
+            assert x[0] == 3
```
Comment (jansel) [inline test/dynamo/test_repros.py:1765]: Can you add a few more test cases:
```
# should get constant propped away (no assert in the graph)
assert x.dtype == torch.float32
assert True
# has message
assert assert x[0] == 3, "foo"
# has message with format
assert assert x[0] == 3, f"foo {x[0]}"
```

Don't need to handle the second two, but we should run correctly on them.

File: torch/_dynamo/config.py [hunk 2]
```diff
@@ -31,7 +31,7 @@
 log_file_name = None
 
 # Verbose will print full stack traces on warnings and errors
-verbose = False
+verbose = True
```
Comment (jansel) [inline torch/_dynamo/config.py:34]: !

File: torch/_dynamo/symbolic_convert.py [hunk 3]
```diff
@@ -125,6 +126,41 @@ def generic_jump(truth_fn: typing.Callable, push: bool):
     def inner(self: "InstructionTranslatorBase", inst: Instruction):
         value: VariableTracker = self.pop()
         self.output.guards.update(value.guards)
+        if config.rewrite_assert_with_torch_assert:
```
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:129]: Pull this into a helper method.

File: torch/_dynamo/symbolic_convert.py [hunk 4]
```diff
@@ -125,6 +126,41 @@ def generic_jump(truth_fn: typing.Callable, push: bool):
...
+            # 22 RAISE_VARARGS           1
+            #
+            # We find this sequence and rewrite them as torch._assert
+
+            # DETECT POP_JUMP_IF_TRUE
+            if truth_fn is operator.truth and not push:
+                current_instruction_pointer = self.instruction_pointer
+                inst = self.instructions[current_instruction_pointer]
+                # Detect LOAD_ASSERTION_ERROR
+                if inst.opname == "LOAD_ASSERTION_ERROR":
```
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:143]: This bytecode was added in 3.9, so might not work in 3.8/3.7

File: torch/_dynamo/symbolic_convert.py [hunk 5]
```diff
@@ -125,6 +126,41 @@ def generic_jump(truth_fn: typing.Callable, push: bool):
...
+                    current_instruction_pointer += 1
+                    if current_instruction_pointer < len(self.instructions):
+                        inst_for_raise = self.instructions[current_instruction_pointer]
+                        # DETECT RAISE_VARARGS
+                        if inst_for_raise.opname == "RAISE_VARARGS":
```
Comment (jansel) [inline torch/_dynamo/symbolic_convert.py:148]: If this were a helper function, you could return rather than deeply nesting if statements.

**Review by jansel — CHANGES_REQUESTED** [review]

File: test/dynamo/test_repros.py (hunk 1 above)
Comment (tugsbayasgalan) [reply test/dynamo/test_repros.py:1765]: will do

File: torch/_dynamo/config.py (hunk 2 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/config.py:34]: oops my bad

File: torch/_dynamo/symbolic_convert.py (hunk 4 above)
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:143]: hmm why is this nested in generic jump instead of a top level

`def LOAD_ASSERTION_ERROR(self, inst)` ?

Comment (suo) [reply torch/_dynamo/symbolic_convert.py:143]: yeah, in particular fbcode uses 3.8, where it looks unfortunately more generic (LOAD_GLOBAL of AssertionError). So I would expect to see a Python version check and different handling here.

File: torch/_dynamo/symbolic_convert.py [hunk 6]
```diff
@@ -125,6 +126,41 @@ def generic_jump(truth_fn: typing.Callable, push: bool):
...
+                            assert_error_msg = ConstantVariable(value="assertion failure")
```
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:149]: This is not sound - the "rhs" logic of the comma for an assert can have anything in it - including logic and function calls. You need to either:

1) Trace it
2) Disallow asserts with anything except strings in that field

```
def f(x): 
    try:
        print("X is", x)
        assert(x < 3), f(1)
        return x
    except:
        pass

f(3)
```

**Review by voznesenskym — CHANGES_REQUESTED** [review]
Please add more test and more strictness to the right hand side logic

File: torch/_dynamo/symbolic_convert.py (hunk 4 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py:143]: Because whenever we see JUMP_INSTRUCTION, we want to look ahead by 2 instructions to see these three form as assertion. So it is little annoying to handle it per instruction.

File: torch/_dynamo/symbolic_convert.py (hunk 6 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py:149]: Sounds good. I think string is good enough for us now.

File: torch/_dynamo/symbolic_convert.py (hunk 3 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py]: done

File: torch/_dynamo/symbolic_convert.py (hunk 4 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py]: done

File: torch/_dynamo/symbolic_convert.py (hunk 5 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py]: done

File: test/dynamo/test_repros.py (hunk 1 above)
Comment (tugsbayasgalan) [reply test/dynamo/test_repros.py]: done

File: torch/_dynamo/symbolic_convert.py (hunk 3 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py]: done

File: torch/_dynamo/symbolic_convert.py (hunk 4 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py]: done

**Comment by tugsbayasgalan** [comment]
Addressed all the comments :) @jansel @suo @voznesenskym

File: torch/_dynamo/symbolic_convert.py [hunk 7]
```diff
@@ -121,10 +122,103 @@ def impl(self: "InstructionTranslatorBase", inst: Instruction):
     return impl
 
 
+def _is_assert_statement(
```
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:125]: Name looks like a naive check, actually has side effects (LOAD_CONST, push, etc). Rename please.

File: torch/_dynamo/symbolic_convert.py [hunk 8]
```diff
@@ -121,10 +122,103 @@ def impl(self: "InstructionTranslatorBase", inst: Instruction):
...
+        if inst.opname != "LOAD_GLOBAL" and inst.arg != 0:
+            return False
+    else:
+        if inst.opname != "LOAD_ASSERTION_ERROR":
+            return False
+
+    current_instruction_pointer += 1
+
+    if current_instruction_pointer >= len(self.instructions):
+        return False
```
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:161]: How can this be?

File: torch/_dynamo/symbolic_convert.py [hunk 9]
```diff
@@ -121,10 +122,103 @@ def impl(self: "InstructionTranslatorBase", inst: Instruction):
...
+            return False
+
+        # CALL_FUNCTION should be followed by RAISE_VARARGS
+        current_instruction_pointer += 1
+        if current_instruction_pointer >= len(self.instructions):
+            return False
+        inst = self.instructions[current_instruction_pointer]
+
+    if inst.opname != "RAISE_VARARGS":
+        return False
```
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:188]: Does this mean we are in LOAD_ASSERTION_ERROR but not LOAD_CONST or RAISE_VARARGS? How is that possible? Let's assert if that's the case?

File: torch/_dynamo/symbolic_convert.py [hunk 10]
```diff
@@ -121,10 +122,103 @@ def impl(self: "InstructionTranslatorBase", inst: Instruction):
...
+    if current_instruction_pointer >= len(self.instructions):
+        return False
+
+    inst = self.instructions[current_instruction_pointer]
+    has_error_msg = False
+    # DETECT RAISE_VARARGS or LOAD CONST
+    if inst.opname == "LOAD_CONST":
+        assert isinstance(
+            inst.argval, str
+        ), f"Assert rhs must be str to be rewritten, but found {type(inst.argval)}"
```
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:169]: This will make all asserts that are like that failures, instead of just graph breaking as they do today, right? This feels like it would regress dynamo logic.

**Review by voznesenskym — CHANGES_REQUESTED** [review]
Not sure we handle the asserts correctly here yet

File: torch/_dynamo/symbolic_convert.py (hunk 8 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py:161]: I am just returning false because there is no way this sequence is going to be assertion as LOAD_ASSERTION_ERROR or LOAD_GLBAL be the last instruction. If we return false here, torchdynamo will switch to default execution path to handle it.

File: torch/_dynamo/symbolic_convert.py (hunk 9 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py:188]: yeah it can be assert.

Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py:188]: i guess there is no weird case where you do LOAD_ASSERTION_ERROR and don't RAISE_VARARGS.

**Comment by tugsbayasgalan** [comment]
Can you elaborate on what else feels wrong besides the comments you left? Or is it good once I handle your comments?

File: torch/_dynamo/symbolic_convert.py (hunk 9 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py:188]: I will just return false because this helper function probably shouldn't error out. If I return false, it will switch back to original dynamo code path, hence no regression.

File: torch/_dynamo/symbolic_convert.py (hunk 8 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py:161]: This is just my attempt of being careful lol

File: torch/_dynamo/symbolic_convert.py (hunk 7 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/symbolic_convert.py:125]: Sounds good. Does `_detect_and_normalize_assert_statement` sound good?

File: torch/_dynamo/config.py [hunk 11]
```diff
@@ -87,6 +87,9 @@
 # if an exception is encountered
 replay_record_enabled = False
 
+# Rewrite assert statement in python with torch._assert
+rewrite_assert_with_torch_assert = False
```
Comment (jansel) [inline torch/_dynamo/config.py:91]: I'd be ok turning this on by default.  It would also be good to turn it on and run CI to see if that shakes out any more issues.

Comment (tugsbayasgalan) [reply torch/_dynamo/config.py:91]: Ok let me flip it and run the CI

Comment (tugsbayasgalan) [reply torch/_dynamo/config.py]: Just did

**Comment by voznesenskym** [comment]
> Can you elaborate on what else feels wrong besides the comments you left? Or is it good once I handle your comments?

Just the comments :)

**Comment by voznesenskym** [comment]
Can you run benchmarks with this diff? 

`time python benchmarks/dynamo/torchbench.py  --accuracy --backend inductor --train`

**Review by jansel — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Comment by DanilBaibak** [comment]
More details about the problem can be found here - [D41332298](https://www.internalfb.com/diff/D41332298)

**Comment by malfet** [comment]
@pytorchbot merge

Manually re-landing as internal tweaks has been made
